fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ProductAnalysisMapper.kt

```diff
@@ -0,0 +1,67 @@
+package org.mozilla.fenix.shopping.middleware
+
+import mozilla.components.browser.engine.gecko.shopping.GeckoProductAnalysis
+import mozilla.components.browser.engine.gecko.shopping.Highlight
+import mozilla.components.concept.engine.shopping.ProductAnalysis
+import org.mozilla.fenix.shopping.store.ReviewQualityCheckState
+import org.mozilla.fenix.shopping.store.ReviewQualityCheckState.OptedIn.ProductReviewState
+
+/**
+ * Maps [ProductAnalysis] to [ProductReviewState].
+ */
+fun ProductAnalysis?.toProductReviewState(): ProductReviewState =
+    if (this == null) {
+        ProductReviewState.Error
+    } else {
+        when (this) {
+            is GeckoProductAnalysis -> toProductReview()
+            else -> ProductReviewState.Error
+        }
+    }
+
+private fun GeckoProductAnalysis.toProductReview(): ProductReviewState =
+    if (productId == null) {
+        ProductReviewState.Error
+    } else {
+        val mappedRating = adjustedRating.toFloatOrNull()
+        val mappedGrade = grade?.toGrade()
+        val mappedHighlights = highlights?.toHighlights()?.toSortedMap()
+
+        if (mappedGrade == null && mappedRating == null && mappedHighlights == null) {
+            ProductReviewState.NoAnalysisPresent
+        } else {
+            ProductReviewState.AnalysisPresent(
+                productId = productId!!,
+                reviewGrade = mappedGrade,
+                needsAnalysis = needsAnalysis,
+                adjustedRating = mappedRating,
+                productUrl = analysisURL!!,
+                highlights = mappedHighlights,
+            )
+        }
+    }
+
+private fun String.toGrade(): ReviewQualityCheckState.Grade? =
+    try {
+        ReviewQualityCheckState.Grade.valueOf(this)
+    } catch (e: IllegalArgumentException) {
+        null
+    }
+
+private fun Highlight.toHighlights(): Map<ReviewQualityCheckState.HighlightType, List<String>>? {
+    val highlights: Map<ReviewQualityCheckState.HighlightType, List<String>?> = mapOf(
+        ReviewQualityCheckState.HighlightType.QUALITY to quality,
+        ReviewQualityCheckState.HighlightType.PRICE to price,
+        ReviewQualityCheckState.HighlightType.SHIPPING to shipping,
+        ReviewQualityCheckState.HighlightType.COMPETITIVENESS to competitiveness,
+    )
+
+    return highlights.filterValues { it != null }.mapValues { it.value!! }.ifEmpty { null }
```

Comment:
is `mapValues` doing anything here? Is it just changing the value type from `List<String>?` to `List<String>`?

Reply:
Good question, yes exactly, it's basically to assert that the type is infact `List<String>`. Not sure if there's a better way to achieve this?

Reply:
I couldn't think of any when I pulled down the code. We can typecast things, but that's still a transformation that has to occur in the `mapValues` block, so I'm not immediately aware of an "optimal" way to de-nullify a type without just force unwrapping it.

Reply:
I think it's fine as we know more that the compiler in this case that value is not null when `mapValues` lamda is triggered 👍